packages/toolpad-app/src/toolpad/DomLoader.tsx

```diff
@@ -330,8 +409,30 @@ export default function DomProvider({ appId, children }: DomContextProps) {
     saveError: null,
     savedDom: dom,
     dom,
+    pendingHistoryDom: dom,
+    undo: [],
+    redo: [],
   });
-  const api = React.useMemo(() => createDomApi(dispatch), []);
+
+  const scheduleHistoryUpdate = React.useMemo(
```

Comment:
Would this work? It could be interesting to rely less on lodash.

```tsx
const scheduleHistoryUpdate = useDebouncedHandler((action: DomAction) => {
  if (action) {
    dispatch({ type: 'DOM_UPDATE_HISTORY' });
  }
}, 500)
```

Reply:
> It could be interesting to rely less on lodash.

It's something I noticed through the codebase - we have many methods written by ourselvs that are part of lodash suite, why do we want to keep doing that rather than use what lodash provides?
There are at least a few issues with this approach:
1. We reimplement something that we can already use
2. We don't cover those with specs (which is done by library)
3. We don't have any documentation, which makes it harder to even know that those things exist in our codebase

If anything I'd think that relying more on libraries that provide common functionality is a good thing, so we can focus on feature development or other more important things 🤔 
Lodash for common utilities and something like [ahooks](https://ahooks.js.org/) for all the different hooks that we have and we could remove quite a big chunk of utilities from our codebase

> Also, in which case would action be falsy? The types don't seem to reflect that behavior

I just checked again - and it seems I no longer get dispatch without action, maybe it's related to the previous use of throttle function 🤔 for some reason `dispatch` was called without the action before.

I just cleaned this up

Reply:
> It's something I noticed through the codebase - we have many methods written by ourselvs that are part of lodash suite, why do we want to keep doing that rather than use what lodash provides?

1. Because `lodash` doesn't provide what we need in an immutable environment where we rely on referential equality to optimize our code, example:

    ```tsx
    const x = { foo: 'bar' }

    x === _.omit({foo:'bar'}, 'hello')
    ```
2. Because `lodash` is full of methods that we now, in 2022 have perfectly fine native javascript constructs for. `_.map`, `_.filter`,...
3. Because `lodash` contains methods that make code hard to read, what does `_.isEmpty` do exactly? Does it error when not pased an array? Why is there an `_.isString`? These clutter the codebase, and make it harder to read.


> If anything I'd think that relying more on libraries that provide common functionality is a good thing, so we can focus on feature development or other more important things

I'm not against relying on libraries, I am against relying on large kitchensink libraries that clutter the codebase with things that are better expressed in native javascript. It is contagious, if you ever want to extract a module from the codebase, it will now have to drag along this huge library that you just need 1 method from.

Reply:
> Because lodash doesn't provide what we need in an immutable environment where we rely on referential equality to optimize our code, example:

Is there no alternative library that works with immutable data structure?

> Because lodash is full of methods that we now, in 2022 have perfectly fine native javascript constructs for. _.map, _.filter,...

I don't see a problem with that - we DONT have to use these methods, there is tons of methods that we can still use from lodash

> Because lodash contains methods that make code hard to read, what does _.isEmpty do exactly? Does it error when not pased an array? Why is there an _.isString? These clutter the codebase, and make it harder to read.

Honestly never had an issue with that, maybe easier to speak about concrete examples, but if something is unclear opening up documentation usually sorted it all out. I'm not sure how having our own undocumented methods makes this less of a problem?

> I'm not against relying on libraries, I am against relying on large kitchensink libraries that clutter the codebase with things that are better expressed in native javascript, it is contagious.

I'm not proposing to use lodash method for native alternatives, I'm proposing not to reinvent exact same methods that do not have native alternatives and have them sit in codebase without a tests or proper documentation

Reply:
I usually don't use `lodash` and use `ramda` instead (it's supposed to be immutable), but actually even when I do, the functions that I use `ramda` for could probably easily be written by myself... So I think it's not a bad approach to avoid using these libraries and using our own implementations instead to save on bundle size, a `debounce` function isn't difficult to have ourselves either.

About some of these functions not being documented - I didn't know about them either, that's why I used lodash in my proposal, but I guess that's what code reviews are for? :) I guess `useDebouncedHandler` would be the choice here.

Also debouncing/throttling seem to be slightly different things, I think throttling always runs the function immediately the first time it's called? So we probably want to use debouncing in this case.

Reply:
> could probably easily be written by myself... 

IMO easily written should not be condition of whether it's worth rewriting it in our codebase or not. It's one thing to copy/paste a snippet, have to maintain it, etc.. we should also then ensure that all such code is properly tested, we pull any updates from the source (if there are any) as well as have the documented properly.

> So I think it's not a bad approach to avoid using these libraries and using our own implementations instead to save on bundle size

I'm of an opposition opinion here - we should avoid prematurely optimising things. Especially when we copy same snippets to our code base we still have to include that code in the bundle. And with a treeshaking importing parts from library shouldn't be such a big deal.

> About some of these functions not being documented - I didn't know about them either, that's why I used lodash in my proposal, but I guess that's what code reviews are for?

While it can prevent such things from happening, I'd argue that reviews should be used for that. It's also very likely that the person knowing about the util method might also miss the different implementation or won't be even part of review in some case.

> I think throttling always runs the function immediately the first time it's called? So we probably want to use debouncing in this case.

It depends on the implementation - lodash allows to control how throttle is called and whether on leading or trailing edge of the call (or could be both). The main difference between throttle and debounce is that thorttle acts as a "rate limiter" where it allows executing function at max times per specified duration, i.e. calling function every 100ms with a throttle of 500ms for 2000ms should lead to 4 function calls. Where debounce waits for specified DURATION to pass after last function call and call the function once. So in the case of every 100ms function call with 500ms duration after 2000ms it should trigger function ONCE at 2500ms point (or something close to that)

Reply:
I think those are valid points too, I guess we just have to pick a standard and go with it? As long as we're consistent I don't think either of these approaches would be bad right now, and we can fix issues as we run into them?

The only point I'm not sure about is what we can do to be more aware of certain `utils` we already have that could be useful. At least some comments clarifying how the utils are supposed to be used would probably be helpful though!